### PR TITLE
Fix crash when a given airloop serves more than one air terminal in the same zone

### LINKED ISSUE #7518: Crash when the same air loop serves more than one terminal unit in the same zone
Issue overview
--------------
This issue is a fork from #7511. The crash there is in SystemReports::InitEnergyReports. It's an array bounds error because `AirLoopNum` is zero for one of the air terminal inlet nodes. The problem was traced back to poor logic in SimAirServingZones::InitAirLoops. The defect file has two terminal units in zone 3-1 which are both served by the same airloop. This *should* be a supported configuration.

### Details
Some additional details for this issue (if relevant):
 - Platform Win10, 64-bit
 - Version of EnergyPlus v9.2.0-44e30c353a

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file attached: [test-v9.2.idf.txt](https://github.com/NREL/EnergyPlus/files/3639081/test-v9.2.idf.txt)
 - [x] Ticket added to Pivotal for defect (development team task)
 - [x] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by mjwitte**
Pull request overview
---------------------
 - Fixes #7518 
 - Fixes `SimAirServingZones::InitAirLoops` which sorts out which airloop number is associated with each air terminal inlet node number. These connections are used later in the simulation. Before this fix, the second (and later) air terminals in the same zone served by the same air loop were never assigned and airloopnum, so it remained zero which caused an array bounds error.
 - No diffs expected.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions: 
*The defect file runs to completion now without crashing (in a debug build).*
 - [ ] ~~If any defect files are updated to a more recent version, upload new versions here or on DevSupport~~
 - [ ] ~~If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label~~
 - [ ] ~~If structural output changes, add to output rules file and add OutputChange label~~
 - [ ] ~~If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies~~

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] ~~If feature, test running new feature, try creative ways to break it~~
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] ~~Verify IDF naming conventions and styles, memos and notes and defaults~~
 - [ ] ~~If new idf included, locally check the err file and other outputs~~

### CODE REVIEW

File: tst/EnergyPlus/unit/SimAirServingZones.unit.cc [hunk 1]
```diff
@@ -630,4 +635,938 @@ TEST_F(EnergyPlusFixture, GetAirPathData_ControllerLockout2)
     EXPECT_TRUE(PrimaryAirSystem(1).CanBeLockedOutByEcono(2));
 }
 
+TEST_F(EnergyPlusFixture, InitAirLoops_1AirLoop2ADU)
```
Comment (mjwitte) [inline tst/EnergyPlus/unit/SimAirServingZones.unit.cc:638]: This unit test has 2 ADUs in separate zones served by a single airloop

File: tst/EnergyPlus/unit/SimAirServingZones.unit.cc [hunk 2]
```diff
@@ -630,4 +635,938 @@ TEST_F(EnergyPlusFixture, GetAirPathData_ControllerLockout2)
...
+    SimAirServingZones::InitAirLoops(true);
+    EXPECT_TRUE(compare_err_stream(""));
+    ASSERT_FALSE(ErrorsFound);
+    // And finally, all of this gymnastics just to check if the airloopnums get set correctly
+    // For this test, both ADUs should be connected airloop 1 which is the only one here
+    EXPECT_EQ(DataZoneEquipment::ZoneEquipConfig(1).InletNodeAirLoopNum(1), 1);
+    EXPECT_EQ(DataZoneEquipment::ZoneEquipConfig(2).InletNodeAirLoopNum(1), 1);
+}
+
+TEST_F(EnergyPlusFixture, InitAirLoops_2AirLoop2ADU)
```
Comment (mjwitte) [inline tst/EnergyPlus/unit/SimAirServingZones.unit.cc:807]: This one has 2 ADUs in separate zones with 2 airloops. Each airloop serves one ADU.

File: tst/EnergyPlus/unit/SimAirServingZones.unit.cc [hunk 3]
```diff
@@ -630,4 +635,938 @@ TEST_F(EnergyPlusFixture, GetAirPathData_ControllerLockout2)
...
+    SimAirServingZones::InitAirLoops(true);
+    EXPECT_TRUE(compare_err_stream(""));
+    ASSERT_FALSE(ErrorsFound);
+    // And finally, all of this gymnastics just to check if the airloopnums get set correctly
+    // For this test, each ADUs should be connected to a different airloop
+    EXPECT_EQ(DataZoneEquipment::ZoneEquipConfig(1).InletNodeAirLoopNum(1), 1);
+    EXPECT_EQ(DataZoneEquipment::ZoneEquipConfig(2).InletNodeAirLoopNum(1), 2);
+}
+
+TEST_F(EnergyPlusFixture, InitAirLoops_2AirLoop3ADUa)
```
Comment (mjwitte) [inline tst/EnergyPlus/unit/SimAirServingZones.unit.cc:1039]: This one adds a 3rd ADU, so zone 2 has 2 ADUs, airloop 1 serves 2 ADUs in different zones, airloop 2 serves the other ADU in zone 2, this one will pass as-is.

File: tst/EnergyPlus/unit/SimAirServingZones.unit.cc [hunk 4]
```diff
@@ -630,4 +635,938 @@ TEST_F(EnergyPlusFixture, GetAirPathData_ControllerLockout2)
...
+    EXPECT_TRUE(compare_err_stream(""));
+    ASSERT_FALSE(ErrorsFound);
+    // And finally, all of this gymnastics just to check if the airloopnums get set correctly
+    // For this test, each ADU 1-1 and 2-2 should be connected to airloop 1, and ADU 2-1 to airloop 1
+    EXPECT_EQ(DataZoneEquipment::ZoneEquipConfig(1).InletNodeAirLoopNum(1), 1);
+    EXPECT_EQ(DataZoneEquipment::ZoneEquipConfig(2).InletNodeAirLoopNum(1), 2);
+    EXPECT_EQ(DataZoneEquipment::ZoneEquipConfig(2).InletNodeAirLoopNum(2), 1);
+}
+
+TEST_F(EnergyPlusFixture, InitAirLoops_2AirLoop3ADUb)
```
Comment (mjwitte) [inline tst/EnergyPlus/unit/SimAirServingZones.unit.cc:1305]: This one is the raison-d'être for this fix: airloop 1 serves one ADU in zone 1, airloop2 serves two ADUs in zone 2. This should fail with current code.

**Review by mjwitte — COMMENTED** [review]
Units tests to show failure. After CI completes, will add the fix.

File: src/EnergyPlus/SimAirServingZones.cc [hunk 5]
```diff
@@ -1770,27 +1770,25 @@ namespace SimAirServingZones {
                                 // and the supply air path outlet
                                 if (SupplyAirPath(SupAirPathNum).OutletNode(SupAirPathOutNodeNum) ==
                                     ZoneEquipConfig(CtrlZoneNum).AirDistUnitCool(ZoneInNum).InNode) {
-                                    if (FindNumberInList(CtrlZoneNum, CtrlZoneNumsCool, NumZonesCool) == 0) {
```
Comment (mjwitte) [inline src/EnergyPlus/SimAirServingZones.cc:1773]: Back in the old days (before a zone could be served by more than one airloop) this test was relevant, but now if there's a matching node, this block should execute.  "Zones" cooled or heated are really air terminal units now. The matching nodes should be a unique connection in the whole scheme of things. This change looks big, but it's just removing this if and it's ending brace.

File: src/EnergyPlus/SimAirServingZones.cc [hunk 6]
```diff
@@ -1808,23 +1806,21 @@ namespace SimAirServingZones {
                                 // BEGIN HEATING: If we don't get a match, check for a heating match
                                 if (SupplyAirPath(SupAirPathNum).OutletNode(SupAirPathOutNodeNum) ==
                                     ZoneEquipConfig(CtrlZoneNum).AirDistUnitHeat(ZoneInNum).InNode) {
-                                    if (FindNumberInList(CtrlZoneNum, CtrlZoneNumsHeat, NumZonesHeat) == 0) {
```
Comment (mjwitte) [inline src/EnergyPlus/SimAirServingZones.cc:1811]: Same change again here in the heating block, remove the if.

File: src/EnergyPlus/SimAirServingZones.cc (hunk 5 above)
Comment (Myoldmopar) [reply src/EnergyPlus/SimAirServingZones.cc:1773]: Change doesn't look so big when you ignore the whitespace changes: https://github.com/NREL/EnergyPlus/pull/7519/files?w=1

File: tst/EnergyPlus/unit/SimAirServingZones.unit.cc (hunk 4 above)
Comment (Myoldmopar) [reply tst/EnergyPlus/unit/SimAirServingZones.unit.cc:1305]: This development process was top-notch.  Adding the unit test and seeing it fail, then fixing the code around it, is just great.  Thanks @mjwitte!

**Review by Myoldmopar — APPROVED** [review]

**Pull request merged**